### PR TITLE
Close #3588 Allow newly enabled module config to be overridden.

### LINKED ISSUE #3588: Settings in az_mail/config/quickstart no longer apply.
## Problem/Motivation
The settings in the `az_mail` module are not being applied when the module is enabled.

### Describe the bug

` [notice] Disallowing override of customized configuration smtp.settings.`

In modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php

The `canOverride` method incorrectly marks configuration that does not have a snapshot yet as `FALSE`.

https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php#L274-L288

### To Reproduce

1. Enable `az_mail` module
2. Go to `/admin/config/system/smtp`
3. See settings are not applied.

## Proposed resolution

Return `TRUE` in `canOverride` in https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php

Something like.
```
    // Guard against missing items.
    $snap = (!empty($snap)) ? $snap : [];
    $active = (!empty($active)) ? $active : [];

    // Not relevant for user roles. Permissions created dynamically.
    if (strpos($name, 'user.role.') !== 0) {
      // Prune cache_metadata if present, to not consider it for diffs.
      $active = $this->trimNestedKey($active, 'cache_metadata');
      $snap = $this->trimNestedKey($snap, 'cache_metadata');
      // If $snap is empty, then we can't compare it to $active.
      // In this case, we can assume that the configuration is not customized
      // because it is not present in the snapshot.
      // This means that the module (config provider) is not yet installed.
      if (empty($snap)) {
        return TRUE;
      }
      // Diff active config and snapshot of module to check for customization.
      $diff = $differ->diff($active, $snap);
      // Overrides only allowed if no changes in diff.
      if (!$this->diffIsEmpty($diff) && !empty($active)) {
        return FALSE;
      }
```

### PR DESCRIPTION
**Pull request by trackleft**
## Description
Not sure if this is correct, but it does allow me to override the smtp.settings via az_mail.

## Related issues
Close #3588 

## How to test
1. Enable az_mail module
2. Go to /admin/config/system/smtp
3. See settings are applied.

Additionally, check the build log for any anomalies.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

### CODE REVIEW

File: modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php [hunk 1]
```diff
@@ -274,11 +273,19 @@ protected function canOverride($name) {
     // Guard against missing items.
     $snap = (!empty($snap)) ? $snap : [];
     $active = (!empty($active)) ? $active : [];
+
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php:276]:
```suggestion
```

**Comment by joeparsons** [comment]
Tested locally with lando/drush.

Before:
```
# lando drush en -y az_mail
The following module(s) will be enabled: az_mail, smtp

 // Do you want to continue?: yes.                                                                                      

 [notice] Disallowing override of customized configuration smtp.settings.
 [success] Successfully enabled: az_mail, smtp
```

After:
```
# lando drush en -y az_mail
The following module(s) will be enabled: az_mail, smtp

 // Do you want to continue?: yes.                                                                                      

 [success] Successfully enabled: az_mail, smtp
```
👍🏻

**Review by joeparsons — APPROVED** [review]
Seemed to work in my local testing.

Was hoping it might also resolve these problems but I think they're caused by something else:
```
 [notice] Could not validate dependencies of override antibot.settings.
 [notice] Could not validate dependencies of override seckit.settings.
```

**Comment by joeparsons** [comment]
Confirmed this issue also affects 2.10.x.

**Comment by tadean** [comment]
It won't let me make a commit suggestion because it's not one of the lines involved in the PR, but maybe something like this below line 271 instead?
```
      // StorageInterface::read() returns FALSE if the config does not exist.
      // In this case, we can assume that the configuration is not customized
      // because it is not present in the snapshot, likely because the module
      // is newly installed.
      if ($snap === FALSE) {
        return TRUE;
      }
 ```
 StorageInterface seems to use `FALSE` to distinguish between non-existent config and empty config.

File: modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php [hunk 2]
```diff
@@ -262,14 +262,20 @@ protected function canOverride($name) {
...
     if (!empty($owner[1])) {
       $snapshot_storage = $this->getConfigSnapshotStorage(ConfigSyncSnapshotterInterface::CONFIG_SNAPSHOT_SET, $owner[0], $owner[1]);
       $snap = $snapshot_storage->read($name);
+      // StorageInterface::read() returns FALSE if the config does not exist.
+      // In this case, we can assume that the configuration is not customized
+      // because it is not present in the snapshot, likely because the module
+      // is newly installed.
+      if ($snap === FALSE) {
+        return TRUE;
+      }
```
Comment (trackleft) [inline modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php:278]: I've heard of an `isNew` property on config objects before in this ticket: https://www.drupal.org/project/config_sync/issues/2996039

I wonder if that could be used here.

Comment (joeparsons) [reply modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php:278]: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21StorableConfigBase.php/function/StorableConfigBase%3A%3AisNew/11.x

**Review by tadean — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]

**Pull request merged**